File: leetcode/438.找到字符串中所有字母异位词/438.找到字符串中所有字母异位词.go
package main

import "fmt"

func findAnagrams(s string, p string) []int {
	need := map[int32]int{}
	window := map[int32]int{}
	for _, i := range p{
		need[i] += 1
		window[i] = 0
	}
	left := 0
	right := 0
	valid := 0
	var result []int
	for right < len(s) {
		c := int32(s[right])
		right += 1
		if _, ok := need[c]; ok {
			window[c] += 1
			if window[c] == need[c] {
				valid += 1
			}
		}
		for right-left >= len(p) {
			if valid == len(need) {
				result = append(result, left)
			}
			d := int32(s[left])
			left += 1
			if _, ok := need[d]; ok {
				if window[d] == need[d] {
					valid -= 1
				}
				window[d] -= 1
			}
		}
	}
	return result
}
func main() {
	s := "cbaebabacd"
	p := "abc"
	result := findAnagrams(s, p)
	fmt.Println(result)
}

File: leetcode/438.找到字符串中所有字母异位词/438.找到字符串中所有字母异位词.py
from typing import List


class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        need, window = {}, {}
        for i in p:
            need[i] = need.get(i, 0) + 1
            window[i] = 0

        left, right = 0, 0
        valid = 0
        result = list()
        # 扩大窗口，对需求的字符计数
        while right < len(s):
            c = s[right]
            right += 1
            if c in need:
                window[c] += 1
                if window[c] == need[c]:
                    valid += 1
            # 缩小窗口，窗口长度大于需求字符串长度时对窗口内容进行判断，满足时添加初始索引即 窗口左区间
            while right - left >= len(p):
                if valid == len(need):
                    result.append(left)
                d = s[left]
                left += 1
                if d in need:
                    if window[d] == need[d]:
                        valid -= 1
                    window[d] -= 1

        return result


if __name__ == '__main__':
    s = "cbaebabacd"
    p = "abc"
    result = Solution().findAnagrams(s, p)
    print(result)
